token/program-2022/src/proof.rs

```diff
@@ -25,3 +25,31 @@ pub fn decode_proof_instruction_context<T: Pod + ZkProofData<U>, U: Pod>(
         .map(ZkProofData::context_data)
         .ok_or(ProgramError::InvalidInstructionData)
 }
+
+/// An `i8` type guaranteed to be non-zero.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct NonZeroI8(i8);
+impl TryFrom<i8> for NonZeroI8 {
+    type Error = ProgramError;
+    fn try_from(n: i8) -> Result<Self, Self::Error> {
+        if n == 0 {
+            Err(ProgramError::InvalidArgument)
+        } else {
+            Ok(Self(n))
+        }
+    }
+}
+impl From<NonZeroI8> for i8 {
+    fn from(n: NonZeroI8) -> Self {
+        n.0
+    }
+}
+
+/// A proof location type meant to be used for arguments to instruction constructors.
+#[derive(Clone, Copy)]
+pub enum ProofLocation<'a, T> {
+    /// The proof is included in the same transaction of a corresponding token-2022 instruction.
+    Instruction(NonZeroI8, &'a T),
```

Comment:
nit so that it's totally clear where to find it
```suggestion
    InstructionOffset(NonZeroI8, &'a T),
```

Or maybe even better, *don't* have this take an i8 value, and instead expose `get_instruction_offset()` function which returns 1 for `InstructionOffset` and 0 for `ContextStateAccount`, and that way you can remove all of the repeated validation logic.

Reply:
I changed this to `InstructionOffset`. I did think about removing the instruction offset, but there could be an application where the proof instruction is 2 or 3 places out from the token instruction. In this case, they would call the `inner_configure_account()` constructor with a custom instruction offset, so it might make sense to leave the offset field in there.